kimchi/src/folding/mod.rs

```diff
@@ -0,0 +1,319 @@
+        }
+    }
+
+    fn witness_ext(&self, i: usize, side: bool) -> &Vec<G::ScalarField> {
+        match side {
+            true => self.a.1.witness.witness_ext(i),
+            false => self.b.1.witness.witness_ext(i),
+        }
+    }
+
+    fn index(&self, i: &GateType) -> &Vec<G::ScalarField> {
+        self.index.get(i).unwrap()
+    }
+
+    fn coefficient(&self, i: usize) -> &Vec<G::ScalarField> {
+        &self.coefficients[i]
+    }
+
+    fn shift(&self) -> &Vec<G::ScalarField> {
+        self.shift
+    }
+
+    fn lagrange_basis(&self, i: &i32) -> Vec<G::ScalarField> {
+        assert!(i >= &0);
+        let zero = G::ScalarField::zero();
+        let mut basis = vec![zero; self.len];
+        basis[*i as usize] = G::ScalarField::one();
+        basis
+    }
+
+    fn zero_vec(&self) -> Vec<G::ScalarField> {
+        let zero = G::ScalarField::zero();
+        vec![zero; self.len]
+    }
+}
+pub trait Sponge<G: CommitmentCurve> {
+    fn challenge(absorbe: &PolyComm<G>) -> G::ScalarField;
+}
+
+pub struct FoldingScheme<G: CommitmentCurve, S: Sponge<G>, C: SRS<G>> {
+    _sponge: PhantomData<S>,
+    // original_exp: Expr<ConstantExpr<G::ScalarField>>,
+    expression: IntegratedFoldingExpr<G::ScalarField>,
+    index: HashMap<GateType, Vec<G::ScalarField>>,
+    coefficients: Vec<Vec<G::ScalarField>>,
+    shift: Vec<G::ScalarField>,
+    srs: Arc<C>,
+    domain: Radix2EvaluationDomain<G::ScalarField>,
+    zero_commitment: PolyComm<G>,
+    zero_vec: Vec<G::ScalarField>,
+}
+impl<G: CommitmentCurve, S: Sponge<G>, C: SRS<G>> FoldingScheme<G, S, C> {
+    pub fn new<F: Fn(&FeatureFlag) -> bool>(
+        expression: &Expr<G::ScalarField>,
+        index: HashMap<GateType, Vec<G::ScalarField>>,
+        coefficients: Vec<Vec<G::ScalarField>>,
+        srs: Arc<C>,
+        domain: Radix2EvaluationDomain<G::ScalarField>,
+        flag_resolver: &F,
+    ) -> Self {
+        let expression = folding_expression(expression, flag_resolver);
+        let shift = domain.elements().collect();
+        let zero: G::ScalarField = G::ScalarField::zero();
+        let evals = std::iter::repeat(zero).take(domain.size()).collect();
+        let zero_vec_evals = Evaluations::from_vec_and_domain(evals, domain);
+        let zero_commitment = srs.commit_evaluations_non_hiding(domain, &zero_vec_evals);
+        let zero_vec = zero_vec_evals.evals;
+        Self {
+            _sponge: PhantomData,
+            expression,
+            index,
+            coefficients,
+            shift,
+            srs,
+            domain,
+            zero_vec,
+            zero_commitment,
+        }
+    }
+
+    pub fn fold_instance_witness_pair<I, W, A, B>(
+        &self,
+        a: A,
+        b: B,
+    ) -> (RelaxedInstance<G, I>, RelaxedWitness<G, W>, PolyComm<G>)
+    where
+        I: InstanceTrait<G>,
+        W: WitnessTrait<G>,
+        A: RelaxablePair<G, I, W>,
+        B: RelaxablePair<G, I, W>,
```

Comment:
Is there a value in assuming `A != B` here?

Reply:
It allows you to fold for example `(RelaxedInstance<G, I>, RelaxedWitness<G, W>)` with `(I,W)`.
Basically, you can provide an instance, witness or pair that could or not be relaxed and fold it with another one that could also be relaxed or not, and it will be relaxed internally if needed.
For IVC you usually have at each step a relaxed pair of accumulated computation and a non-relaxed pair for the new step.

Reply:
Getting back to the previous comment on struct/trait hierarchy: I think it should be `I: RelaxableInstance, W: RelaxableWitness`, since not all the instances are relaxable. It makes sense, as you say, that you can merge relaxed pair with a standard (but relaxable) pair, but the naming is a bit confusing. 

I'll resolve this particular discussion.